.config/doom/config.org
#+TITLE: Doom Emacs
#+author: Jacob Hilker
#+date: 2022-03-19
#+hugo_base_dir: ../../
#+hugo_type: config
#+hugo_draft: false
#+hugo_custom_front_matter: :mermaid true
#+hugo_section: emacs
#+hugo_bundle: doom
#+export_file_name: index

This is my doom emacs literate configuration, containing everything necessary for my setup.

* Why Emacs?
Emacs allows me to combine multiple workflows I have with one program, whereas with another editor I would need different programs for each workflow I have.
#+begin_mermaid
flowchart LR
    todo[Task Management] --- editor[Code Editor]
    todo[Task Management] --- Email
#+end_mermaid

* Modules
:PROPERTIES:
:header-args:emacs-lisp: :tangle init.el
:END:
My init is pretty basic.
#+begin_src emacs-lisp +n 1 :noweb noexport
;;; init.el -*- lexical-binding: t; -*-

;; This file controls what Doom modules are enabled and what order they load
;; in. Remember to run 'doom sync' after modifying it!

;; NOTE Press 'SPC h d h' (or 'C-h d h' for non-vim users) to access Doom's
;;      documentation. There you'll find a link to Doom's Module Index where all
;;      of our modules are listed, including what flags they support.

;; NOTE Move your cursor over a module's name (or its flags) and press 'K' (or
;;      'C-c c k' for non-vim users) to view its documentation. This works on
;;      flags as well (those symbols that start with a plus).
;;
;;      Alternatively, press 'gd' (or 'C-c c d') on a module to browse its
;;      directory (for easy access to its source code).

(doom! :input
       ;;chinese
       ;;japanese
       ;;layout            ; auie,ctsrnm is the superior home row

       :completion
       company           ; the ultimate code completion backend
       ;;helm              ; the *other* search engine for love and life
       ;;ido               ; the other *other* search engine...
       ;;ivy               ; a search engine for love and life
       (vertico           ; the search engine of the future
        +icons)

       :ui
       ;;deft              ; notational velocity for Emacs
       doom              ; what makes DOOM look the way it does
       doom-dashboard    ; a nifty splash screen for Emacs
       doom-quit         ; DOOM quit-message prompts when you quit Emacs
       (emoji +unicode +github)  ; 🙂
       hl-todo           ; highlight TODO/FIXME/NOTE/DEPRECATED/HACK/REVIEW
       hydra
       ;;indent-guides     ; highlighted indent columns
       ;;ligatures         ; ligatures and symbols to make your code pretty again
       ;;minimap           ; show a map of the code on the side
       modeline          ; snazzy, Atom-inspired modeline, plus API
       ;;nav-flash         ; blink cursor line after big motions
       ;;neotree           ; a project drawer, like NERDTree for vim
       ophints           ; highlight the region an operation acts on
       (popup +defaults)   ; tame sudden yet inevitable temporary windows
       ;;tabs              ; a tab bar for Emacs
       treemacs          ; a project drawer, like neotree but cooler
       unicode           ; extended unicode support for various languages
       vc-gutter         ; vcs diff in the fringe
       vi-tilde-fringe   ; fringe tildes to mark beyond EOB
       ;;window-select     ; visually switch windows
       workspaces        ; tab emulation, persistence & separate workspaces
       zen               ; distraction-free coding or writing

       :editor
       (evil +everywhere); come to the dark side, we have cookies
       file-templates    ; auto-snippets for empty files
       fold              ; (nigh) universal code folding
       (format +onsave)  ; automated prettiness
       ;;god               ; run Emacs commands without modifier keys
       ;;lispy             ; vim for lisp, for people who don't like vim
       ;;multiple-cursors  ; editing in many places at once
       ;;objed             ; text object editing for the innocent
       ;;parinfer          ; turn lisp into python, sort of
       ;;rotate-text       ; cycle region at point between text candidates
       snippets          ; my elves. They type so I don't have to
       word-wrap         ; soft wrapping with language-aware indent

       :emacs
       (dired             ; making dired pretty [functional]
        +icons
        +ranger)
       electric          ; smarter, keyword-based electric-indent
       ;;ibuffer         ; interactive buffer management
       undo              ; persistent, smarter undo for your inevitable mistakes
       vc                ; version-control and Emacs, sitting in a tree

       :term
       ;;eshell            ; the elisp shell that works everywhere
       ;;shell             ; simple shell REPL for Emacs
       ;;term              ; basic terminal emulator for Emacs
       ;;vterm             ; the best terminal emulation in Emacs

       :checkers
       syntax              ; tasing you for every semicolon you forget
       ;;(spell +flyspell) ; tasing you for misspelling mispelling
       ;;grammar           ; tasing grammar mistake every you make

       :tools
       ;;ansible
       ;;biblio            ; Writes a PhD for you (citation needed)
       ;;debugger          ; FIXME stepping through code, to help you add bugs
       ;;direnv
       ;;docker
       ;;editorconfig      ; let someone else argue about tabs vs spaces
       ;;ein               ; tame Jupyter notebooks with emacs
       (eval +overlay)     ; run code, run (also, repls)
       ;;gist              ; interacting with github gists
       lookup              ; navigate your code and its documentation
       ;;lsp               ; M-x vscode
       magit             ; a git porcelain for Emacs
       ;;make              ; run make tasks from Emacs
       (pass +auth)              ; password manager for nerds
       pdf               ; pdf enhancements
       ;;prodigy           ; FIXME managing external services & code builders
       ;;rgb               ; creating color strings
       ;;taskrunner        ; taskrunner for all your projects
       ;;terraform         ; infrastructure as code
       ;;tmux              ; an API for interacting with tmux
       ;;upload            ; map local to remote projects via ssh/ftp

       :os
       (:if IS-MAC macos)  ; improve compatibility with macOS
       ;;tty               ; improve the terminal Emacs experience

       :lang
       ;;agda              ; types of types of types of types...
       ;;beancount         ; mind the GAAP
       ;;(cc +lsp)         ; C > C++ == 1
       ;;clojure           ; java with a lisp
       ;;common-lisp       ; if you've seen one lisp, you've seen them all
       ;;coq               ; proofs-as-programs
       ;;crystal           ; ruby at the speed of c
       ;;csharp            ; unity, .NET, and mono shenanigans
       ;;data              ; config/data formats
       ;;(dart +flutter)   ; paint ui and not much else
       ;;dhall
       ;;elixir            ; erlang done right
       ;;elm               ; care for a cup of TEA?
       emacs-lisp        ; drown in parentheses
       ;;erlang            ; an elegant language for a more civilized age
       ;;ess               ; emacs speaks statistics
       ;;factor
       ;;faust             ; dsp, but you get to keep your soul
       ;;fortran           ; in FORTRAN, GOD is REAL (unless declared INTEGER)
       ;;fsharp            ; ML stands for Microsoft's Language
       ;;fstar             ; (dependent) types and (monadic) effects and Z3
       ;;gdscript          ; the language you waited for
       ;;(go +lsp)         ; the hipster dialect
       ;;(haskell +lsp)    ; a language that's lazier than I am
       ;;hy                ; readability of scheme w/ speed of python
       ;;idris             ; a language you can depend on
       ;;json              ; At least it ain't XML
       ;;(java +lsp)       ; the poster child for carpal tunnel syndrome
       ;;javascript        ; all(hope(abandon(ye(who(enter(here))))))
       ;;julia             ; a better, faster MATLAB
       ;;kotlin            ; a better, slicker Java(Script)
       ;;latex             ; writing papers in Emacs has never been so fun
       ;;lean              ; for folks with too much to prove
       ;;ledger            ; be audit you can be
       ;;lua               ; one-based indices? one-based indices
       markdown          ; writing docs for people to ignore
       ;;nim               ; python + lisp at the speed of c
       ;;nix               ; I hereby declare "nix geht mehr!"
       ;;ocaml             ; an objective camel
       org               ; organize your plain life in plain text
       ;;php               ; perl's insecure younger brother
       ;;plantuml          ; diagrams for confusing people more
       ;;purescript        ; javascript, but functional
       ;;python            ; beautiful is better than ugly
       ;;qt                ; the 'cutest' gui framework ever
       ;;racket            ; a DSL for DSLs
       ;;raku              ; the artist formerly known as perl6
       ;;rest              ; Emacs as a REST client
       ;;rst               ; ReST in peace
       ;;(ruby +rails)     ; 1.step {|i| p "Ruby is #{i.even? ? 'love' : 'life'}"}
       ;;rust              ; Fe2O3.unwrap().unwrap().unwrap().unwrap()
       ;;scala             ; java, but good
       ;;(scheme +guile)   ; a fully conniving family of lisps
       sh                ; she sells {ba,z,fi}sh shells on the C xor
       ;;sml
       ;;solidity          ; do you need a blockchain? No.
       ;;swift             ; who asked for emoji variables?
       ;;terra             ; Earth and Moon in alignment for performance.
       ;;web               ; the tubes
       ;;yaml              ; JSON, but readable
       ;;zig               ; C, but simpler

       :email
       ;;(mu4e +org +gmail)
       ;;notmuch
       ;;(wanderlust +gmail)

       :app
       ;;calendar
       ;;emms
       ;;everywhere        ; *leave* Emacs!? You must be joking
       (irc)               ; how neckbeards socialize
       (rss +org)        ; emacs as an RSS reader
       ;;twitter           ; twitter client https://twitter.com/vnought

       :config
       literate
       (default +bindings +smartparens))
#+end_src
* Rudimentary Configuration
** Personal Info
#+begin_src emacs-lisp
(setq user-full-name "Jacob Hilker"
      user-mail-address "jacob.hilker2@gmail.com"
      auth-sources '("~/.authinfo.gpg"))
#+end_src

* UI
I like to use gruvbox, along with a customized build of Iosevka I've patched for nerd-fonts.
#+begin_src emacs-lisp
(setq doom-theme 'doom-gruvbox
      doom-font (font-spec :family "Josevka" :size 14))
#+end_src



* Applications
** IRC (Circe)
#+begin_src emacs-lisp

#+end_src


emacs/.config/doom/config.org
#+title: Doom Emacs Config



* UI
** Theme
I currently like using gruvbox as my theme.
#+begin_src elisp
(setq doom-theme 'doom-gruvbox)
#+end_src

** Fonts
I am using Iosevka Nerd Font as my font.
#+begin_src elisp
(setq doom-font (font-spec :family "Iosevka Nerd Font" :size 15)
      doom-variable-pitch-font (font-spec :family "Iosevka Aile" :size 15)
      doom-big-font (font-spec :family "Iosevka Nerd Font" :size 24))
#+end_src



* Application
** Elfeed
#+begin_src elisp
(after! elfeed
  (elfeed-goodies/setup)
  (setq elfeed-search-filter "@1-week-ago +unread"
      elfeed-goodies/feed-source-column-width 30
      elfeed-goodies/tag-column-width 16))
#+end_src


.config/qtile/config.py
import os
import re
import socket
import subprocess
from typing import List  # noqa: F401

from libqtile import layout, bar, widget, hook
from libqtile.lazy import lazy
from libqtile.config import Match, Screen, Group, DropDown, ScratchPad, KeyChord
from libqtile.config import EzKey as Key, EzClick as Click, EzDrag as Drag

from qtile_extras import widget

from theme import fonts
from theme import gruvbox as colors

mod = "mod4"
terminal = "alacritty"
browser = "firefox"

# TODO Fix matching

groups = [Group("1", layout='monadtall'), ## Browser
          Group("2", layout='monadtall'), ## Emacs/Dev
          Group("3", layout='monadtall'), ## Terminal

          Group("4", layout='monadtall'), ## IRC/Discord
          Group("5", layout='monadtall'), ## Steam/Games
          Group("6", layout='monadtall'), ## Music
          Group("7", layout='max'), ## Graphics
          Group("8", layout='max'),
          Group("9", layout='max')]

keys = [
    Key("M-h", lazy.layout.left(), desc="Move focus to left"),
    Key("M-l", lazy.layout.right(), desc="Move focus to right"),
    Key("M-j", lazy.layout.down(), desc="Move focus down"),
    Key("M-k", lazy.layout.up(), desc="Move focus up"),
    Key("M-S-h", lazy.layout.shuffle_left(), desc="Move window to the left"),
    Key("M-S-l", lazy.layout.shuffle_right(), desc="Move window to the right"),
    Key("M-S-j", lazy.layout.shuffle_down(), desc="Move window down"),
    Key("M-S-k", lazy.layout.shuffle_up(), desc="Move window up"),
    Key("M-C-h", lazy.layout.grow_left(), desc="Grow window to the left"),
    Key("M-C-l", lazy.layout.grow_right(), desc="Grow window to the right"),
    Key("M-C-j", lazy.layout.grow_down(), desc="Grow window down"),
    Key("M-C-k", lazy.layout.grow_up(), desc="Grow window up"),
    Key("M-C-<equal>", lazy.layout.grow()),
    Key("M-C-<minus>", lazy.layout.shrink()),
    Key("M-C-0", lazy.layout.reset()),

    Key("M-S-q", lazy.window.kill(), desc="Kill focused window"),
    Key("M-S-r", lazy.restart(), desc="Restart Qtile"),
    Key("M-S-p", lazy.spawn("rofi -show powermenu -theme-str 'window { height: 55%;}' "), desc="Manage machine power state"),
    Key("M-f", lazy.window.toggle_fullscreen(), desc="Toggle fullscreen"),
    Key("M-S-f", lazy.window.toggle_floating(), desc="Toggle floating layout"),

    Key("M-<Tab>", lazy.next_layout()),
    Key("<XF86AudioLowerVolume>", lazy.spawn("pactl set-sink-volume @DEFAULT_SINK@ -5%")),
    Key("<XF86AudioRaiseVolume>", lazy.spawn("pactl set-sink-volume @DEFAULT_SINK@ +5%")),
    Key("<XF86AudioMute>", lazy.spawn("pactl set-sink-mute @DEFAULT_SINK@ toggle")),

    Key("M-r", lazy.spawn("rofi -show drun")),
    Key("M-s", lazy.spawn("alacritty -t 'ncspot' -e 'ncspot'")),
    Key("M-<Return>", lazy.spawn(terminal), desc="Launch terminal"),
    Key("M-S-<Return>", lazy.spawn(browser), desc="Open Firefox"),
    
    KeyChord([mod], "e", [
        Key("c", lazy.spawn("emacsclient -cs 'doom -e '(erc)'"), desc="Spawn erc with emacs client"),
        Key("e", lazy.spawn("emacsclient -cs 'doom' -a 'emacs'"), desc="Spawn emacs client"),
        Key("n", lazy.spawn("emacsclient -cs 'doom' -e '(elfeed)'"), desc="Spawn elfeed with emacs client"),
    ], mode="Emacs Apps"),

    KeyChord([mod, "shift"], "e", [
        Key("d", lazy.spawn("emacsclient -cs 'doom' -a 'emacs'"), desc="Spawn Doom Emacs client"),
        Key("j", lazy.spawn("emacsclient -cs 'jmacs' -a 'emacs --with-profile=jmacs'"), desc="Spawn Jmacs client"),
    ], mode="Emacs Profiles"),


        

    

    Key("C-<grave>", lazy.spawn("dunstctl close")),
]

for group in groups:
    keys.extend([
      Key("M-{}".format(group.name), lazy.group[group.name].toscreen(), desc="Switch to group {}".format(group.name)),
      Key("M-S-{}".format(group.name), lazy.window.togroup(group.name), desc="Move focused window to group {}".format(group.name)) 
    ])

layout_theme = {
    "margin": 10,
    "border_focus": colors['purple'],
    "border_normal": colors['bg'],
    "border_width": 2
}

layouts = [
    layout.MonadTall(**layout_theme),
    layout.Max(**layout_theme),
]

floating_layout = layout.Floating(float_rules=[
    # Run the utility of `xprop` to see the wm class and name of an X client.
    *layout.Floating.default_float_rules,
    Match(wm_class='confirmreset'),  # gitk
    Match(wm_class='makebranch'),  # gitk
    Match(wm_class='maketag'),  # gitk
    Match(wm_class='ssh-askpass'),  # ssh-askpass
    Match(title='branchdialog'),  # gitk
    Match(title='pinentry'),  # GPG key password entry
    Match(wm_class='pinentry-gtk-2'), 
], **layout_theme)

widget_defaults = dict(
    font=fonts['text'],
    fontsize=14,
    padding=3,
    background = colors['bg'],
    foreground = colors['fg'],
    theme_path = "/usr/share/icons/Paper/16x16/panel/"
)

extension_defaults = widget_defaults.copy()

def draw_arrow_right(bg,fg,font_size=33):
    "Creates a textbox widget with a right-pointing arrow."
    return widget.TextBox(text="",
                          padding=0,
                          fontsize=font_size,
                          background=bg,
                          foreground=fg)

def draw_arrow_left(bg,fg,font_size=33):
    "Creates a textbox widget with a right-pointing arrow."
    return widget.TextBox(text="",
                          padding=0,
                          fontsize=font_size,
                          background=bg,
                          foreground=fg)

mainbar = bar.Bar([
    widget.Sep(linewidth=0,
               padding=6),
    widget.GroupBox(disable_drag=True,
                    block_highlight_text_color=colors['fg'],
                    active=colors['fg'],),
    draw_arrow_right(colors['blue'],
                     colors['bg']),
    widget.TextBox(text="",
                   font=fonts['material'],
                   fontsize = 12,
                   background=colors['blue']),
    widget.Clock(format="%H:%M - %a %d %b",
                 background=colors['blue']),

    draw_arrow_right(colors['purple'],
                     colors['blue']),
    widget.CurrentLayout(background=colors['purple']),
    draw_arrow_right(colors['bg'],
                     colors['purple']),
    #widget.Spacer(),
    widget.WindowName(),
    draw_arrow_left(colors['bg'],
                    colors['orange']),
    widget.WiFiIcon(background=colors['orange'],
                    active_colour = colors['fg'],
                    inactive_colour = colors['gray'],
                    padding=7,
                    foreground=colors['fg']),


    draw_arrow_left(colors['orange'],
                    colors['purple']),
    widget.TextBox(text="",
                   font = fonts['material'],
                   fontsize=16,
                   background=colors['purple']),

    widget.Bluetooth(hci="/dev_90_7A_58_A6_A0_0A",
                     background=colors['purple']),


    draw_arrow_left(colors['purple'],
                    colors['green']),


    widget.GenPollText(update_interval=None,
                       func=lambda: subprocess.check_output(os.path.expanduser("~/.dotfiles/.config/qtile/scripts/volicon.sh")).decode('utf-8'),
                       font=fonts['material'],
                       fontsize=16,
                       background=colors['green']),

    widget.GenPollText(update_interval=None,
                       func=lambda: subprocess.check_output(os.path.expanduser("~/.dotfiles/.config/qtile/scripts/printvol.sh")).decode('utf-8'),
                       background=colors['green']),

], 30)

screens = [
    Screen(top = mainbar),
]

@hook.subscribe.startup_once
def autostart():
    home = os.path.expanduser('~')
    subprocess.call([home + '/.config/qtile/scripts/autostart.sh'])
